extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java

```diff
@@ -0,0 +1,747 @@
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response postTable(
+      @PathParam("dbSchema") String dbSchema,
+      @PathParam("name") String name,
+      TableSpec spec,
+      @QueryParam("action") String actionParam,
+      @QueryParam("version") long version,
+      @Context final HttpServletRequest req
+  )
+  {
+    final PostAction action;
+    if (actionParam == null) {
+      action = PostAction.NEW;
+    } else {
+      action = PostAction.valueOf(StringUtils.toUpperCase(actionParam));
+      if (action == null) {
+        return Actions.badRequest(
+            Actions.INVALID,
+            StringUtils.format(
+                "Not a valid action: [%s]. Valid actions are new, ifNew, replace, force",
+                actionParam
+            )
+        );
+      }
+    }
+    TableId tableId = TableId.of(dbSchema, name);
+    Response response = authorizeTable(tableId, spec, req);
+    if (response != null) {
+      return response;
+    }
+    TableMetadata table = TableMetadata.newTable(tableId, spec);
+    try {
+      catalog.validate(table);
+    }
+    catch (IAE e) {
+      return Actions.badRequest(Actions.INVALID, e.getMessage());
+    }
+
+    switch (action) {
+      case NEW:
+        return insertTableSpec(table, false);
+      case IFNEW:
+        return insertTableSpec(table, true);
+      case REPLACE:
+        return updateTableSpec(table, version);
+      case FORCE:
+        return addOrUpdateTableSpec(table);
+      default:
+        throw new ISE("Unknown action.");
+    }
+  }
+
+  private Response authorizeTable(TableId tableId, TableSpec spec, final HttpServletRequest req)
+  {
+    // Druid has a fixed set of schemas. Ensure the one provided is valid.
+    Pair<Response, SchemaSpec> result = validateSchema(tableId.schema());
+    if (result.lhs != null) {
+      return result.lhs;
+    }
+    SchemaSpec schema = result.rhs;
+
+    // The schema has to be one that allows table definitions.
+    if (!schema.writable()) {
+      return Actions.badRequest(
+          Actions.INVALID,
+          StringUtils.format("Cannot modify schema %s", tableId.schema())
+      );
+    }
+
+    // Table name can't be blank or have spaces
+    if (Strings.isNullOrEmpty(tableId.name())) {
+      return Actions.badRequest(Actions.INVALID, "Table name is required");
+    }
+    if (!tableId.name().equals(tableId.name().trim())) {
+      return Actions.badRequest(Actions.INVALID, "Table name cannot start or end with spaces");
+    }
+
+    // The user has to have permission to modify the table.
+    try {
+      catalog.authorizer().authorizeTable(schema, tableId.name(), Action.WRITE, req);
+    }
+    catch (ForbiddenException e) {
+      return Actions.forbidden(e);
+    }
+
+    // Validate the spec, if provided.
+    if (spec != null) {
+
+      // The given table spec has to be valid for the given schema.
+      if (Strings.isNullOrEmpty(spec.type())) {
```

Comment:
Should a TableSpec ever not have a type? If not, maybe we should move this validation to the TableSpec class itself?